File: leetcode/22.括号生成/22.括号生成.go
package main

import (
	"fmt"
	"strings"
)

func generateParenthesis(n int) []string {
	res := []string{}
	if n == 0 {
		return res
	}
	track := []string{}
	var backtrack func(left int, right int)
	backtrack = func(left int, right int) {
		if left > right {
			return
		}
		if left < 0 || right < 0 {
			return
		}
		if left == 0 && right == 0 {
			res = append(res, strings.Join(track, ""))
			return
		}
		track = append(track, "(")
		backtrack(left-1, right)
		track = track[:len(track)-1]

		track = append(track, ")")
		backtrack(left, right-1)
		track = track[:len(track)-1]
	}
	backtrack(n, n)
	return res
}

func main() {
	n := 3
	result := generateParenthesis(n)
	fmt.Println(result)
}

File: leetcode/22.括号生成/22.括号生成.py
from typing import List

"""
括号性质：
    1. 一个合法括号的左括号数量一定等于右括号数量
    2. 一个合法括号的字串的左括号数量都大于活等于右括号数量
问题分解：
计算n对括号的组合
    组合（回溯）
    合法括号（根据性质）
"""


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        track = []
        if n == 0:
            return res

        # left 左括号数量，right 右括号数量
        def backtrack(left, right):
            # 若左括号剩下的多，说明不合法
            if right < left:
                return
            # 括号数量小于0
            if left < 0 or right < 0:
                return
            # 合法数量都为0
            if left == 0 and right == 0:
                res.append(''.join(track))
                return
            # 遍历选择，放括号
            # 放左括号
            track.append('(')
            backtrack(left - 1, right)
            track.pop()

            # 放右括号
            track.append(')')
            backtrack(left, right - 1)
            track.pop()

        backtrack(n, n)
        return res


if __name__ == '__main__':
    n = 3
    result = Solution().generateParenthesis(n)
    print(result)
